lighthouse-core/report/html/renderer/element-screenshot-renderer.js

```diff
@@ -0,0 +1,256 @@
+    const elementRectCenter = getRectCenterPoint(elementRectInScreenshotCoords);
+
+    // Try to center clipped region.
+    const screenshotLeftVisibleEdge = clamp(
+      elementRectCenter.x - elementPreviewSizeInScreenshotCoords.width / 2,
+      0, screenshotSize.width - elementPreviewSizeInScreenshotCoords.width
+    );
+    const screenshotTopVisisbleEdge = clamp(
+      elementRectCenter.y - elementPreviewSizeInScreenshotCoords.height / 2,
+      0, screenshotSize.height - elementPreviewSizeInScreenshotCoords.height
+    );
+
+    return {
+      screenshot: {
+        left: screenshotLeftVisibleEdge,
+        top: screenshotTopVisisbleEdge,
+      },
+      clip: {
+        left: elementRectInScreenshotCoords.left - screenshotLeftVisibleEdge,
+        top: elementRectInScreenshotCoords.top - screenshotTopVisisbleEdge,
+      },
+    };
+  }
+
+  /**
+   * @param {DOM} dom
+   * @param {HTMLElement} maskEl
+   * @param {{left: number, top: number}} positionClip
+   * @param {LH.Artifacts.Rect} elementRectInScreenshotCoords
+   * @param {Size} elementPreviewSizeInScreenshotCoords
+   */
+  static renderClipPath(dom, maskEl, positionClip,
+      elementRectInScreenshotCoords, elementPreviewSizeInScreenshotCoords) {
+    const clipPathEl = dom.find('clipPath', maskEl);
+    const clipId = `clip-${Util.getUniqueSuffix()}`;
+    clipPathEl.id = clipId;
+    maskEl.style.clipPath = `url(#${clipId})`;
+
+    // Normalize values between 0-1.
+    const top = positionClip.top / elementPreviewSizeInScreenshotCoords.height;
+    const bottom =
+      top + elementRectInScreenshotCoords.height / elementPreviewSizeInScreenshotCoords.height;
+    const left = positionClip.left / elementPreviewSizeInScreenshotCoords.width;
+    const right =
+      left + elementRectInScreenshotCoords.width / elementPreviewSizeInScreenshotCoords.width;
+
+    const polygonsPoints = [
+      `0,0             1,0            1,${top}          0,${top}`,
+      `0,${bottom}     1,${bottom}    1,1               0,1`,
+      `0,${top}        ${left},${top} ${left},${bottom} 0,${bottom}`,
+      `${right},${top} 1,${top}       1,${bottom}       ${right},${bottom}`,
+    ];
+    for (const points of polygonsPoints) {
+      clipPathEl.append(dom.createElementNS(
+        'http://www.w3.org/2000/svg', 'polygon', undefined, {points}));
+    }
+  }
+
+  /**
+   * @param {DOM} dom
+   * @param {LH.Artifacts.FullPageScreenshot} fullPageScreenshot
+   */
+  static createBackgroundImageStyle(dom, fullPageScreenshot) {
+    const styleEl = dom.createElement('style');
+    styleEl.id = 'full-page-screenshot-style';
+    styleEl.innerText = `
+      .lh-element-screenshot__image {
+        background-image: url(${fullPageScreenshot.data})
+      }`;
+    return styleEl;
+  }
+
+  /**
+   * Installs the lightbox elements and wires up click listeners to all .lh-element-screenshot elements.
+   * Should only be called exactly once per report render.
```

Comment:
> should we enforce the exactly once part?

This is the kind of thing that could definitely become an issue in the future, e.g. loading more than one report in the same viewer or devtools or etc. Is there some kind of check it could do (like I believe an older version did?) to check if it has already been called?